Fix 32bit support and add workflow for 32bits testing

come-nc opened the pull request:
Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>


* Resolves: #36072

## Summary

Add 32bits to CI and fix errors

## TODO

- [x] Have a working 32bits PHPUnit CI step
- [x] Refine which tests it runs or not
- [x] Fix errors
- [x] Refine when this CI step runs or not
- [ ] Backport CI step (and fixes if needed) to stable25

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

come-nc committed:
Add workflow for 32bits testing

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Revert "remove 32-bit workarounds"

This reverts commit dd8774389e21b59c07882580356d51de018fe867.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
```
There were 2 failures:

1) Test\User\UserTest::testSetQuota
Expectation failed for method name is "setUserValue" when invoked 1 time(s)
Parameter 3 for invocation OCP\IConfig::setUserValue('foo', 'files', 'quota', '0 B', null) does not match expected value.
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'23 TB'
+'0 B'

/__w/server/server/lib/private/User/User.php:526
/__w/server/server/tests/lib/User/UserTest.php:744

2) Test\User\UserTest::testSetQuotaAddressNoChange
OCP\IConfig::setUserValue('foo', 'files', 'quota', '0 B', null) was not expected to be called.

/__w/server/server/lib/private/User/User.php:526
/__w/server/server/tests/lib/User/UserTest.php:831
```

So it works.

But either we do not have enough tests on quota or they are not run by the selection I’ve put, because I expected type errors as well on file sizes.
We need to add tests specifically for files over 4GB I guess.

There is also a whole bunch of:
```
1) OCA\DAV\Tests\unit\CalDAV\CalDavBackendTest::testCalendarQuery with data set "all" (array(0, 1, 2, 3), array(), array())
ValueError: Epoch doesn't fit in a PHP integer

/__w/server/server/apps/dav/lib/CalDAV/CalDavBackend.php:2738
/__w/server/server/apps/dav/lib/CalDAV/CalDavBackend.php:1209
/__w/server/server/apps/dav/tests/unit/CalDAV/AbstractCalDavBackend.php:221
/__w/server/server/apps/dav/tests/unit/CalDAV/CalDavBackendTest.php:426
```

come-nc committed:
Running all tests on 32bits to see what happens

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Add apcu and ffmpeg to have less skipped tests

Also exclude test groups which rely on a service

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
(rebased)

come-nc committed:
Strong type filesize related methods to ease 32bits problem findings

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Fix psalm errors related to filesizes

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Improve test failure output for RouterTest.php

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Fix bigsize support on 32bits in OC_Helper

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Fix filesize return type error in ViewTest

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Add helper to cast to int|float, fix MemoryInfo on 32bits

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Fix 32bits bigint support in Util/OC_Helper

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Fix a few tests that were using too big int values for 32bits

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
For the `ValueError: Epoch doesn't fit in a PHP integer` errors from dav tests, they are triggered by calling `getTimestamp` on `DateTime` objects for dates after 2038.
We need to decide whether we want to support events after 2038 in dav/calendar or not for 32bits, which would imply retyping all timestamps in there to int|float. Other solution is to document it as known limitation on 32bits and skip those tests on 32bits.
ping @PVince81 @AndyScherzinger

come-nc committed:
Fix computerFileSize when called with a string without unit

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Fix setQuota on User on 32bits

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Allow float sizes in FileInfo for 32bits

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Type sizes as int|float throughout the code base

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Improve error output on failed Router test

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Silence errors of return type mismatch from Sabre for now

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

AndyScherzinger commented:
> support events after 2038

I'd be fine to document it as a known limitation, that 32-bit can't handle dates past that time.

come-nc committed:
Improve behavior with dates on 32bits and fix tests or skip them

We do not support events after 2038 on 32bits but still behave better
 when date range start/end is after 2038.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

ChristophWurst commented:
review was wrongly requested because the PR was set for review already. put it back to draft.

come-nc committed:
Put back string in $includeMounts as it actually does something

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Fix name of skip method for phpunit

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Fix psalm errors

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Fix Sabre connector test

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Psalm easy fixes in apps/files/appinfo/routes.php

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Skip buggy tests for 32bits CI

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Skip tests that cannot succeed on 32bits

For some reason a timestamp in the past also triggers the PHP error
 about not being able to represent it as an int, so skipping that test
 on 32bits.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Only run 32bit CI once a day

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
> * [ ]  Document that events past 2038 (check actual limit) are not supported on 32bit

- [ ] Document that events before EPOCH are not supported either on 32bit

come-nc commented:
CI is now passing, rebased on master, ready for review.
I’ve set the 32bit CI to run once a day at 1:15 am, what do you think @AndyScherzinger ?

come-nc committed:
Document that size may be a float in new class VersionEntity

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

AndyScherzinger commented:
> I’ve set the 32bit CI to run once a day at 1:15 am, what do you think

@come-nc sounds good to me, time and interval wise 👍

artonge commented on lib/private/legacy/OC_Response.php:
```diff
@@ -52,6 +52,19 @@ public static function setContentDispositionHeader($filename, $type = 'attachmen
 	 */
 	public static function setContentLengthHeader($length) {
+		if (PHP_INT_SIZE === 4) {
+			if ($length > PHP_INT_MAX && stripos(PHP_SAPI, 'apache') === 0) {
```
Check if we should do that all the time

come-nc replied:
We checked with @artonge and it should be fine

come-nc commented:
I fixed psalm warning, but drone testing is failing with:
```
173	There was 1 error:
174	
175	1) Test\Files\ViewTest::testLockBasicOperationUnlocksAfterCancelledHook with data set #14 ('filesize', array('dir'), 'dir', null, 1, 1, 1, null, 100)
176	ParseError: syntax error, unexpected token "|", expecting "{"
177	
178	/drone/src/lib/private/Files/View.php:1182
179	/drone/src/lib/private/Files/View.php:415
180	/drone/src/tests/lib/Files/ViewTest.php:2092
```

I have no idea where this comes from, there is no "|" character at the line it’s giving. I’m failing to reproduce locally so far.

artonge commented:
> I have no idea where this comes from, there is no "|" character at the line it’s giving. I’m failing to reproduce locally so far.

I think it refers to that: https://github.com/nextcloud/server/blob/39aeecff46748673899b9288f81e98c50a647192/lib/private/Files/Storage/Local.php#L245 (that's the method called by `/drone/src/lib/private/Files/View.php:1182` is think)

Is drone using a PHP version compatible with this syntax ?

come-nc committed:
Removed catch of ValueError as we cannot know if it’s >2038 or <1970

Also fixed numericToString to correctly convert float to int if it fits

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented on lib/public/Util.php:
```diff
@@ -339,25 +339,36 @@ public static function getDefaultEmailAddress(string $user_part): string {
+	 * @return int|float int if it fits, float if it is too big
+	 * @since 26.0.0
+	 */
+	public static function numericToNumber(string|float|int $number): int|float {
```
In the end I’ve left these types and added a cast to string instead, this makes sure the method wil cast for float to int if the value fit in an int

come-nc commented:
> > I have no idea where this comes from, there is no "|" character at the line it’s giving. I’m failing to reproduce locally so far.
> 
> I think it refers to that:
> 
> https://github.com/nextcloud/server/blob/39aeecff46748673899b9288f81e98c50a647192/lib/private/Files/Storage/Local.php#L245
> 
> (that's the method called by `/drone/src/lib/private/Files/View.php:1182` I think)
> Is drone using a PHP version compatible with this syntax ?

No it uses PHP 8.0.26, and I checked on 3v4l, the syntax you link to is accepted.

I rebased on master and pushed fixes to Louis remarks.

come-nc commented:
Extracted the complete stacktrace:
```
[sqlite-php8.0:81] ParseError: syntax error, unexpected token "|", expecting "{" in phar:///usr/local/bin/phpunit/phpunit/Framework/MockObject/MockClass.php(48) : eval()'d code:3
[sqlite-php8.0:82] Stack trace:
[sqlite-php8.0:83] #0 phar:///usr/local/bin/phpunit/phpunit/Framework/MockObject/Generator.php(345): PHPUnit\Framework\MockObject\MockClass->generate()
[sqlite-php8.0:84] #1 phar:///usr/local/bin/phpunit/phpunit/Framework/MockObject/Generator.php(121): PHPUnit\Framework\MockObject\Generator->getObject()
[sqlite-php8.0:85] #2 phar:///usr/local/bin/phpunit/phpunit/Framework/MockObject/Invocation.php(149): PHPUnit\Framework\MockObject\Generator->getMock()
[sqlite-php8.0:86] #3 phar:///usr/local/bin/phpunit/phpunit/Framework/MockObject/InvocationHandler.php(129): PHPUnit\Framework\MockObject\Invocation->generateReturnValue()
[sqlite-php8.0:87] #4 phar:///usr/local/bin/phpunit/phpunit/Framework/MockObject/MockClass.php(48) : eval()'d code(24): PHPUnit\Framework\MockObject\InvocationHandler->invoke()
[sqlite-php8.0:88] #5 /drone/src/lib/private/Files/View.php(1182): Mock_Temporary_b8d464a7->filesize()
[sqlite-php8.0:89] #6 /drone/src/lib/private/Files/View.php(415): OC\Files\View->basicOperation()
[sqlite-php8.0:90] #7 [internal function]: OC\Files\View->filesize()
[sqlite-php8.0:91] #8 /drone/src/tests/lib/Files/ViewTest.php(2093): call_user_func_array()
[sqlite-php8.0:92] #9 phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php(1254): Test\Files\ViewTest->testLockBasicOperationUnlocksAfterCancelledHook()
[sqlite-php8.0:93] #10 phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php(920): PHPUnit\Framework\TestCase->runTest()
[sqlite-php8.0:94] #11 phar:///usr/local/bin/phpunit/phpunit/Framework/TestResult.php(596): PHPUnit\Framework\TestCase->runBare()
[sqlite-php8.0:95] #12 phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php(706): PHPUnit\Framework\TestResult->run()
[sqlite-php8.0:96] #13 phar:///usr/local/bin/phpunit/phpunit/Framework/TestSuite.php(506): PHPUnit\Framework\TestCase->run()
[sqlite-php8.0:97] #14 phar:///usr/local/bin/phpunit/phpunit/Framework/TestSuite.php(506): PHPUnit\Framework\TestSuite->run()
[sqlite-php8.0:98] #15 phar:///usr/local/bin/phpunit/phpunit/Framework/TestSuite.php(506): PHPUnit\Framework\TestSuite->run()
[sqlite-php8.0:99] #16 phar:///usr/local/bin/phpunit/phpunit/Framework/TestSuite.php(506): PHPUnit\Framework\TestSuite->run()
[sqlite-php8.0:100] #17 phar:///usr/local/bin/phpunit/phpunit/TextUI/TestRunner.php(487): PHPUnit\Framework\TestSuite->run()
[sqlite-php8.0:101] #18 phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php(129): PHPUnit\TextUI\TestRunner->run()
[sqlite-php8.0:102] #19 phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php(98): PHPUnit\TextUI\Command->run()
[sqlite-php8.0:103] #20 /usr/local/bin/phpunit(1091): PHPUnit\TextUI\Command::main()
[sqlite-php8.0:104] #21 {main}
```
We may have hit a phpunit bug, not sure.

come-nc commented:
> We may have hit a phpunit bug, not sure.

Confirmed, should be fixed by phpunit update which was published in the docker image.

come-nc commented:
Rebased

nickvergessen commented on lib/public/Files/Storage.php:
```diff
@@ -133,12 +133,9 @@ public function filetype($path);
 	 * @since 6.0.0
 	 */
-	public function filesize($path);
+	public function filesize(string $path): false|int|float;
```
This will break storage providers implementing this interface?

come-nc replied:
Yes but it will ensure they support 32bits.
I can move OCP type changes to phpdoc know that tests are passing if it is preferred.
But sadly it will also mean removing lots of typing from private classes implementing these interfaces, and it is this typing which allowed to catch a lot of 32bits related errors with the tests.

nickvergessen replied:
Yeah I'm drawn apart. Could do a voting/opinion gathering

come-nc replied:
Moved it to phpdoc

nickvergessen commented on lib/public/Files/Storage/IStorage.php:
```diff
@@ -130,12 +130,9 @@ public function filetype($path);
 	 * @since 9.0.0
 	 */
-	public function filesize($path);
+	public function filesize(string $path): false|int|float;
```
This will break storage providers implementing this interface?

nickvergessen commented on lib/public/Files/SimpleFS/ISimpleFile.php:
```diff
@@ -49,7 +49,7 @@ public function getName(): string;
 	 * @since 11.0.0
 	 */
-	public function getSize(): int;
+	public function getSize(): int|float;
```
Needs documentation at least when we assume it does not have apps implementing it

ChristophWurst replied:
Also apps that just work against the interface need to adjust for this.

come-nc replied:
Well this is the only way to not crash on 32bit for files over 2G. So application do have to handle float, but before they had to handle TypeError in the same situation, so I do not think this will bring any difference, if they rely on the var being int they will have a TypeError thrown from their code instead of being thrown from inside OC.

nickvergessen commented on tests/lib/Encryption/UtilTest.php:
```diff
@@ -195,6 +195,7 @@ public function dataTestIsSystemWideMountPoint() {
 
 	/**
 	 * @dataProvider dataTestIsSystemWideMountPoint
+	 * @group BrokenTests
```
Please remove or better rebase on top of https://github.com/nextcloud/server/pull/36504

come-nc replied:
rebased

nickvergessen commented on tests/lib/Route/RouterTest.php:
```diff
@@ -30,12 +30,20 @@
 /**
  * Class RouterTest
  *
+ * @group RoutingWeirdness
```
what

come-nc replied:
This group of tests only pass on drone, not on github CI or locally.
But they only fail when run as part of some group of tests, if you run only this class you have no trouble.
It has to do with include_once and how phpunit handles resetting the state between tests, I **think**.
At some point I had to stop trying to fix them and just skip them as their failure is not related to 32bit.

So I added them to a group that I skip in 32bit CI.

come-nc committed:
Move back typing to phpdoc to keep backward compatibility in OCP

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

nickvergessen commented:
Github is complaining 🙈

come-nc committed:
Respect OCP interface in private classes

Because the parameter type was moved to phpdoc it needs to be removed
 from implementations

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

ChristophWurst reviewed (dismissed):
Breaking changes in the public API and we are past feature freeze

nickvergessen commented on lib/public/Files/FileInfo.php:
```diff
@@ -84,7 +84,7 @@ public function getEtag();
 	 *
 	 * @param bool $includeMounts whether or not to include the size of any sub mounts, since 16.0.0
-	 * @return int
+	 * @return int|float
```
Is it actually breaking, or just making sure everyone is handling the floats that always where returned by 32bit systems.

come-nc replied:
No, it always returned float of 32 bit, the documentation was just lying.

ChristophWurst replied:
Fair enough.

come-nc commented:
The drone samba failure passes locally so false positive I think.
The kerberos one I’m failing to run locally. @icewind1991 Any idea?

come-nc committed:
Show logs on smb-kerberos CI failure

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
Samba kerberos failure actually seems to be #36314

come-nc commented:
/rebase

DaphneMuller commented:
hello @come-nc  
Thank you so much for your work on this pull request! This ticket seems to have the tag 'pending documentation'. 
Is there any chance you could clarify what documentation is missing? Is this for admins or for app developers?

MichaIng commented:
@come-nc 
Why was this never backported to NC25? It is high likely what fixes quotas on NC26 while it is still broken on NC25. It is is overall to much/large, we could also cherry-pick some commits?

come-nc commented:
> @come-nc Why was this never backported to NC25? It is high likely what fixes quotas on NC26 while it is still broken on NC25. It is is overall to much/large, we could also cherry-pick some commits?

Several reasons, mainly this was fixed through strong typing using union type `int|float` which is only available on PHP 8.0 and higher, and thus cannot be used on 25.
So backporting would have meant doing it again using phpdoc instead of real typing.

Also you can see I had to change things in lib/public (OCP), which we tend to avoid on bugfix releases.

MichaIng commented:
Okay, makes sense. Then it will be some detectives work to find the part which fixes quota usage on Nextcloud 25. Or the monkeys task to translate all union types to phpdoc comments for backporting everything 😄. I'll see if I can do that, excluding CI and unrelated stuff.

> Also you can see I had to change things in lib/public (OCP), which we tend to avoid on bugfix releases.

In this cases it fixes bugs, so I think it is necessary. But I also think most of these are in NC25 already (phpdoc).

MichaIng commented:
PR up which backports most of the changes done here: #37877
